### PR TITLE
adding :add_curl_script option that, when checked, let a curl script …

### PR DESCRIPTION
**Pull request by itzhakishahar**
…be added to the event msg (in :curl key).

### CODE REVIEW

File: lib/rack/logstash-writer.rb [hunk 1]
```diff
@@ -5,112 +5,167 @@
...
+    # @param [Hash] env rack env obj
+    # @return [String] e.g. "METHOD name/path/path?query=query protocol"
+    def request_line(env)
+      line = "#{env['REQUEST_METHOD']} #{env['SCRIPT_NAME']}#{env['PATH_INFO']}"
+      line << "?#{env['QUERY_STRING']}" if env['QUERY_STRING'] && !env['QUERY_STRING'].empty?
+      line << " #{env['SERVER_PROTOCOL']}"
     end
 
+    module CURL
+      DEFAULT_ARGUMENT_CONFIGS = [
```
Comment (eranb) [inline lib/rack/logstash-writer.rb]: put every thing inside a function, don't use module

File: lib/rack/logstash-writer.rb [hunk 2]
```diff
@@ -5,112 +5,167 @@
...
+          {pattern: '--user-agent "%s"', env_key: 'HTTP_USER_AGENT'}
+      ]
+
+      # @param [Hash] env as received from Rack
+      # @return [String] of curl unix-command that will imitate the request to server/service
+      def self.gen_curl_by_rack_env(env)
+        curl_base = "curl #{env['rack.url_scheme']}://#{env['SERVER_NAME']}:#{env['SERVER_PORT']}#{env['REQUEST_URI']}"
+        DEFAULT_ARGUMENT_CONFIGS.each_with_object([curl_base]) do |arg_config, cmd_acc|
+          arg_value = arg_config[:env_key] && env[arg_config[:env_key]] || arg_config[:value]
+          arg_value = proc(&arg_config[:proc]).call(arg_value) if arg_config[:proc] rescue arg_value
```
Comment (eranb) [inline lib/rack/logstash-writer.rb]:
```ruby
arg_config[:proc].call(arg_value) # == proc(&arg_config[:proc]).call(arg_value)
```

File: lib/rack/logstash-writer.rb [hunk 3]
```diff
@@ -5,112 +5,167 @@
...
+    # @return [String] e.g. "METHOD name/path/path?query=query protocol"
+    def request_line(env)
+      line = "#{env['REQUEST_METHOD']} #{env['SCRIPT_NAME']}#{env['PATH_INFO']}"
+      line << "?#{env['QUERY_STRING']}" if env['QUERY_STRING'] && !env['QUERY_STRING'].empty?
+      line << " #{env['SERVER_PROTOCOL']}"
     end
 
+    module CURL
+      DEFAULT_ARGUMENT_CONFIGS = [
+          {pattern: '--request %s', rack_env_key: 'REQUEST_METHOD'},
```
Comment (eranb) [inline lib/rack/logstash-writer.rb]: Why you specify each header and don't take all of them

File: lib/rack/logstash-writer.rb (hunk 2 above)
Comment (itzhakishahar) [reply lib/rack/logstash-writer.rb]: key :to_json insted of overly dynamic :proc

File: examples/rails-project/config.ru [hunk 4]
```diff
@@ -3,6 +3,15 @@ require 'rack/logstash-writer'
...
-# use Rack::LogstashWriter, "tcp://localhost:5228" , {} , [*(100..600)]#"tcp://localhost:5228" # #
+use Rack::LogstashWriter, {
+    url: "file:///home/org/Desktop/logsample",
+    statuses: [*(100..600)],
+    body_len: 100,
+    response_headers: {'message' => 'This-is-a-message-man'},
+    request_headers: {'User-agent' => 'ua-nimrod'},
+    body_regex: {service_name: 'service_namev:(.*).*[,]?.*}'},
+    proc: -> env { {service: Dir.pwd.split("/").last} },
+    add_curl_script: true
```
Comment (eranb) [inline examples/rails-project/config.ru]: curl_cmd?

File: examples/regular rack file/config.ru [hunk 5]
```diff
@@ -15,14 +15,17 @@ class JSONServerError
...
-    body_regex: {service_name: 'service_namev:(.*).*[,]?.*}' , proc: Proc.new {|env| {service: Dir.pwd.split("/").last}}}
+use Rack::LogstashWriter, {
+    url: "file:///home/org/Desktop/logsample", # could also be another using protocol, e.g. : "udp://localhost:5228", "tcp://localhost:5228"
+    request_headers: {'head1' => 'head1'},
+    response_headers: {'head1' => 'head1'},
+    statuses: [*(500..600)],
+    body_len: 50,
+    body_regex: {service_name: 'service_namev:(.*).*[,]?.*}'},
+    proc: -> env { {service: Dir.pwd.split("/").last} },
+    add_curl_script: true
```
Comment (eranb) [inline examples/regular rack file/config.ru]: :curl_cmd?

File: README.md [hunk 6]
```diff
@@ -46,7 +46,7 @@ use Rack::LogstashWriter , {
     proc: -> env { {a_key: env[:a_key]} },
     # {optional} [Boolean] add [:curl] to the log with *curl* script to mock the request that led to the event; default: false.
     # aliased as :add_curl_script
-    add_request_script: true
+    add_request_script?: true
```
Comment (eranb) [inline README.md]: add_curl?

File: examples/rails-project/config.ru [hunk 7]
```diff
@@ -11,7 +11,7 @@ use Rack::LogstashWriter, {
     request_headers: {'User-agent' => 'ua-nimrod'},
     body_regex: {service_name: 'service_namev:(.*).*[,]?.*}'},
     proc: -> env { {service: Dir.pwd.split("/").last} },
-    add_curl_script: true
+    add_curl_command?: true
```
Comment (eranb) [inline examples/rails-project/config.ru]: add_curl?

File: examples/regular rack file/config.ru [hunk 8]
```diff
@@ -25,7 +25,7 @@ use Rack::LogstashWriter, {
     body_len: 50,
     body_regex: {service_name: 'service_namev:(.*).*[,]?.*}'},
     proc: -> env { {service: Dir.pwd.split("/").last} },
-    add_curl_script: true
+    add_curl_command?: true
```
Comment (eranb) [inline examples/regular rack file/config.ru]: add_curl?

File: examples/sinatra-project/config.ru [hunk 9]
```diff
@@ -15,7 +15,7 @@ use Rack::LogstashWriter, {
     response_headers: {'message' => 'This-is-a-message-man'},
     request_headers: {'User-agent'=>'ua-nimrod'}, # could also be using another protocol, e.g. : "udp://localhost:5228", "file:///home/org/Desktop/logsample"
     body_regex: {service_name: 'service_namev:(.*).*[,]?.*}' , proc: Proc.new {|env| {service: Dir.pwd.split("/").last}}},
-    add_curl_script: true # will help
+    add_curl_command?: true
```
Comment (eranb) [inline examples/sinatra-project/config.ru]: add_curl?

File: lib/rack/logstash-writer.rb [hunk 10]
```diff
@@ -10,9 +10,11 @@ class LogstashWriter
         response_headers: nil,
         statuses: [*(500..600)],
         body_len: 1000,
-        add_curl_script: false,
-        add_request_script: false
+        add_curl_command?: false,
```
Comment (eranb) [inline lib/rack/logstash-writer.rb]: add_curl?

File: lib/rack/logstash-writer.rb [hunk 11]
```diff
@@ -24,8 +26,8 @@ class LogstashWriter
     # @option options [Fixnum] :body_len: optional, include the first given chars from the body. default 1000
     # @option options [Array] :statuses: optional, send events to log stash only for those statuses. default [*(500..600)]
     # @option options [Proc] :proc: optional, the function will call the proc with the env and call
-    # @option options [Boolean] :add_curl_script: add [:curl] to the log with *curl* unix command string set to mock the request that led to the log
-    # @option options [Boolean] :add_request_script: alias for the :curl: @option
+    # @option options [Boolean] :add_curl_command?: add [:curl] to the log with *curl* unix command string set to mock the request that led to the log
```
Comment (eranb) [inline lib/rack/logstash-writer.rb]: add_curl?

File: lib/rack/logstash-writer.rb [hunk 12]
```diff
@@ -24,8 +26,8 @@ class LogstashWriter
...
+    # @option options [Boolean] :add_request_script?: alias for the :curl: @option
```
Comment (eranb) [inline lib/rack/logstash-writer.rb]: add_curl?

File: lib/rack/logstash-writer.rb [hunk 13]
```diff
@@ -95,7 +97,10 @@ def log(env, status, response_headers, began_at, body)
         end
       end
 
-      data[:curl] = CURL.gen_curl_by_rack_env(env) if @options[:add_curl_script] || @options[:add_request_script]
+      # add curl if users enables it
+      if @options[:add_curl_command?] || @options[:add_request_script]
```
Comment (eranb) [inline lib/rack/logstash-writer.rb]: add_curl?

File: README.md (hunk 6 above)
Comment (itzhakishahar) [reply README.md]: or wanted to have an opt that does not mention the name of the command ("curl").
there is still an option called "add_curl_command?" that is mentioned here as an alias and can be seen in logstash-writer.rb.

**Pull request merged**